### PR TITLE
[CAMEL-11257] First Phase of AS2 Camel Component

### PR DESCRIPTION
**Pull request by punkhorn**
This development phase of the AS2 Component provides support for signed and unsigned transport of EDI message and synchronous signed and unsigned Message Disposition Notification (MDN) reports.

### CODE REVIEW

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/AS2MessageDispositionNotificationEntity.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    private static final String FINAL_RECIPIENT = "Final-Recipient";
+    private static final String ORIGINAL_MESSAGE_ID = "Original-Message-ID";
+    private static final String AS2_DISPOSITION = "Disposition";
+    private static final String FAILURE = "Failure";
+    private static final String ERROR = "Error";
+    private static final String WARNING = "Warning";
+    private static final String RECEIVED_CONTENT_MIC = "Received-content-MIC";
+
+    private String reportingUA;
+    // TODO determine if we need to support this field.
```
Comment (apupier) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/AS2MessageDispositionNotificationEntity.java]: is it a leftover? or there is a specific task for it?

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/ApplicationPkcs7SignatureEntity.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        // Write out mime part headers if this is not the main body of message.
+        if (!isMainBody()) {
+            try (CanonicalOutputStream canonicalOutstream = new CanonicalOutputStream(ncos, AS2Charset.US_ASCII)) {
+
+                HeaderIterator it = headerIterator();
+                while (it.hasNext()) {
+                    Header header = it.nextHeader();
+                    canonicalOutstream.writeln(header.toString());
+                }
+                canonicalOutstream.writeln(); // ensure empty line between
```
Comment (apupier) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/ApplicationPkcs7SignatureEntity.java]:
```
NoCloseOutputStream ncos = new NoCloseOutputStream(outstream);
+        try (CanonicalOutputStream canonicalOutstream = new CanonicalOutputStream(ncos, AS2Charset.US_ASCII)) {
+
+            // Write out mime part headers if this is not the main body of
+            // message.
+            if (!isMainBody()) {
+                HeaderIterator it = headerIterator();
+                while (it.hasNext()) {
+                    Header header = it.nextHeader();
+                    canonicalOutstream.writeln(header.toString());
+                }
+                canonicalOutstream.writeln(); // ensure empty line between
+                                              // headers and body; RFC2046 -
+                                              // 5.1.1
```

seems to be duplicated 5 times, maybe worth refactoring it

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2MicAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.as2.api;
+
+interface Constants {
+    static final String SHA_1_AS2_ALGORITHM_NAME = "sha1";
+    static final String SHA_1_JDK_ALGORITHM_NAME = "SHA-1";
+
+    static final String MD5_AS2_ALGORITHM_NAME = "md5";
```
Comment (davsclaus) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2MicAlgorithm.java]: Is the MD5 algorithm not too weak today?

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ServerConnection.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+import org.apache.http.protocol.ResponseConnControl;
+import org.apache.http.protocol.ResponseContent;
+import org.apache.http.protocol.ResponseDate;
+import org.apache.http.protocol.ResponseServer;
+import org.apache.http.protocol.UriHttpRequestHandlerMapper;
+import org.apache.http.util.Args;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AS2ServerConnection {
```
Comment (davsclaus) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ServerConnection.java]: Hmm we are doing manualy HTTP server code here with threads, loops etc. I wonder if using a HTTP server library is not better? For example undertow

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2SignedDataGenerator.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.as2.api;
+
+import java.security.Key;
+/**
```
Comment (davsclaus) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2SignedDataGenerator.java]: Remove this license header that is a duplicate

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2MicAlgorithm.java (hunk 3 above)
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2MicAlgorithm.java]: Probably but it is cited as supported in RFC4130

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ServerConnection.java (hunk 4 above)
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ServerConnection.java]: I think a full blown HTTP server would be overkill here. The listener and handler threads used here are simple and targeted for what needs to be done to specifically handle a received AS2 message and no more. We are also using the same underlying library, httpcore, that other http base components such as camel-http4 are using which I think adds to maintainability.

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2SignedDataGenerator.java (hunk 5 above)
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2SignedDataGenerator.java]: Sorry @davsclaus I don't follow? There is only one license header here and it is required by the Check Style rule.

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientManager.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            try {
+                MultipartSignedEntity multipartSignedEntity = new MultipartSignedEntity(applicationEDIEntity, gen,
+                        AS2Charset.US_ASCII, AS2TransferEncoding.BASE64, true, null);
+                EntityUtils.setMessageEntity(request, multipartSignedEntity);
+            } catch (Exception e) {
+                throw new HttpException("Failed to sign message", e);
+            }
+            break;
+        case ENCRYPTED:
+            break;
```
[the remark this answers is not in the document]
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientManager.java]: Yes code will be added here when encryption facility is added. I have added comments stetting that per you suggestion.

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/AS2MessageDispositionNotificationEntity.java (hunk 1 above)
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/AS2MessageDispositionNotificationEntity.java]: No. I put this comment in to remind myself to reevaluate whether the MDN gateway field feature is ever being used once the component gets into the field.

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/ApplicationPkcs7SignatureEntity.java (hunk 2 above)
Comment (punkhorn) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/ApplicationPkcs7SignatureEntity.java]: Perhaps. We'll see going forward.

**Comment by davsclaus** [comment]
Okay its squashed and merged to master

@punkhorn can you rebase from master, and continue your work from there, and submit new PRs. As one giant PR is too hard to review.

**Comment by punkhorn** [comment]
@davsclaus will do and understood. Thanks!

**Pull request merged**